File: redak.cpp
/* #ident "$Id: $"
 * @author: rzr@gna.org - rev: $Author: rzr$
 * Copyright: See README file that comes with this distribution
 *****************************************************************************/

#include "config.h"

#include "redak.h"
#include <QtPlugin>
#include <QProcess>


Redak::Redak(QDeclarativeItem *parent)
: QDeclarativeItem(parent)
//: QDeclarativeExtensionPlugin(parent)
{
    FUNCT();
}


void Redak::registerTypes(const char *uri)
{
    FUNCT();
    qDebug()<<uri;
    // Q_ASSERT(uri == QLatin1String("redak"));
    uri="redak";
    qmlRegisterType<Redak>(uri, 1, 1, "Redak");
}


bool Redak::save(QString content, QString filename)
{
    FUNCT();
    bool status = true;
    QUrl url(filename);
    filename = url.path();

#if defined Q_OS_SYMBIAN && !defined Q_WS_SIMULATOR
    filename = filename.mid(1);  //TODO WORAROUND BUG
#endif

    // qDebug()<<"save:" + filename;
    // qDebug()<<"content:" + content;
    QFile file( filename );
    status &= file.open(QIODevice::WriteOnly | QIODevice::Text);
    status &= file.isWritable();

    if ( status ) {
        QTextStream stream( &file );
        stream<<content;
        file.close();
        emit saved();
    }

    if (!status) {
        QString text = "error: io: save: "+filename;
        // qDebug()<< text;

        emit error(QVariant(text));
    }
    return status;
}


QString Redak::load(QString filename)
{
    FUNCT();
    // qDebug()<<"load: " + filename;
    QString content = "";
    bool status = true;

    QUrl url(filename);
    filename = url.path();

#if defined Q_OS_SYMBIAN && !defined Q_WS_SIMULATOR
    filename = filename.mid(1);  //TODO WORAROUND BUG
#endif

    QFile file(filename); //TODO: on dir ?

    if ( file.exists() ) {
        status &= file.open(QIODevice::ReadOnly | QIODevice::Text);
        QTextStream stream( &file );
        content = stream.readAll();
        file.close();
        emit loaded(content);
    } else {
        status &= file.open(QIODevice::WriteOnly | QIODevice::Text);
        content = "# file://" + filename + "\n";
        file.close();
        emit loaded(content);
    }

    if (!status) {
        QString text = "error: io: load: "+filename;
        emit error(QVariant(text));
    }

    return content;
}


QString Redak::process(QString const filename)
{
     QProcess tmp;
     tmp.start(filename);
     tmp.closeWriteChannel();
     tmp.waitForFinished();
     QString content = tmp.readAllStandardOutput();

     return (content);
}


Q_EXPORT_PLUGIN2(Redak, Redak);

//qmlRegisterType<Redak>("Redak", 1, 0, "Redak");

File: redak.py
#! /usr/bin/env python
# #ident "$Id: $"
# @author: rzr@gna.org - rev: $Author: rzr$
# Copyright: See README file that comes with this distribution
################################################################################
# -*- encoding: iso-8859-15 -*-

import sys
import PySide

from PySide import QtCore
from PySide.QtCore import QObject, QUrl, QProcess
from PySide.QtGui import QApplication
from PySide.QtDeclarative import QDeclarativeView,QDeclarativeEngine

from PySide.QtCore import Signal as pyqtSignal
from PySide.QtCore import Slot as pyqtSlot


__package__ = 'redak'
__author__ = 'Philippe Coval (RzR)'
__email__ = 'rzr@gna.org'
__version__ = '0.0.0'


class Redak(QDeclarativeView):

    contentChanged = pyqtSignal(str)


    def __init__(self, parent=None):
	platform = 'common'
	platform = 'desktop'
        src = 'qml/redak/' + platform + '/main.qml'
        super(Redak, self).__init__(parent)
        self.setWindowTitle(__package__)
        self.setSource(QUrl.fromLocalFile(src))
        self.setResizeMode(QDeclarativeView.SizeRootObjectToView)
        self.setGeometry(200, 200, 400, 400)
#       self.showFullScreen()
        root = self.rootObject()
        root.quit.connect( self.quit )
        root.test.connect( self.test )
        self.rootContext().setContextProperty("parentCore", self);
        print "http://bugs.pyside.org/show_bug.cgi?id=1000"
#       self.contentChanged.connect( self.slotContent );
#       self.contentChanged.connect( root.onContentChanged );
#       self.contentChanged.connect( root.handleContent );


    def quit(self):
#       print "quitting"
        self.rootContext().setContextProperty("parentCore", None );
        PySide.QtCore.QCoreApplication.quit()


    def test(self):
#       print "Testing emit"
        content = "content"
        print "Content: " + content
        self.contentChanged.emit( content )
        return content;


    @QtCore.Slot(unicode,result=unicode)
    def process(self,command):
#       print "process"
        content = "content"
        process = QProcess()
        process.start(command)
        process.closeWriteChannel()
        process.waitForFinished()
#       process.closeWriteChannel()
        content = process.readAllStandardOutput()
        print "Content: " + content
        return unicode(content);


    @QtCore.Slot(unicode,result=unicode)
    def load(self,command):
        return "todo"


    @QtCore.Slot(unicode,unicode,result=bool)
    def save(self,content,filename):
        return false


    @QtCore.Slot(str)
    def slotContent(self,content):
        print "#{ slotContent"
        root = self.rootObject()
        root.handleContent(content)
        print content


    @staticmethod
    def main(argv):
        if len(argv) > 1:
            print PySide.QtCore.qVersion()
        app = QApplication(argv)
        view = Redak()
        view.show()
        sys.exit(app.exec_())


if __name__ == "__main__":
    Redak.main(sys.argv)


#eof
